tokio/src/io/poll_evented.rs

```diff
@@ -171,7 +171,15 @@ feature! {
             loop {
                 let evt = ready!(self.registration.poll_read_ready(cx))?;
 
-                let b = &mut *(buf.unfilled_mut() as *mut [std::mem::MaybeUninit<u8>] as *mut [u8]);
+                // SAFETY:
+                //
+                // 1. `MaybeUninit<u8>` has the same memory layout as `u8`.
+                // 2. `*mut [u8] as *mut [MaybeUninit<u8>]` follows the
+                //    [Pointer-to-pointer cast].
+                //
+                // [Pointer-to-pointer cast]:
+                // https://doc.rust-lang.org/1.90.0/reference/expressions/operator-expr.html#r-expr.as.pointer
+                let b = unsafe { &mut *(buf.unfilled_mut() as *mut [std::mem::MaybeUninit<u8>] as *mut [u8]) };
```

Comment:
If we're going to document this cast, we should explain that we trust the IO resource to not read the uninitialized data in this buffer.